File: src/main/scala/com/leetcode/easy/ReverseInteger.scala
package com.leetcode.easy

object ReverseInteger {
  def reverse(x: Int): Int = {
    if (x == Int.MinValue) return 0
    
    var res = 0
    var n = x
    while (n != 0) {
      val unitDigit = n % 10
      if (res < 0 && res < (Int.MinValue - unitDigit) / 10) return 0
      if (res > 0 && res > (Int.MaxValue - unitDigit) / 10) return 0
      res = res * 10 + unitDigit
      n = n / 10
    }
    res
  }
  
  def reverse2(x: Int): Int = {
    if (x > -10 && x < 10) return x
    
    val symbol = if (x < 0) -1 else 1
    val arr = x.toLong.abs.toString.toCharArray
    val reversedArr = reverseArr(arr)
    val reversedStr = String.valueOf(reversedArr)
    val reversedNum = reversedStr.toLong * symbol
    if (reversedNum < Int.MinValue || reversedNum > Int.MaxValue) 0
    else reversedNum.toInt
  }
  
  def reverseArr(arr: Array[Char]): Array[Char] = {
    var i = 0
    var j = arr.size - 1
    while (i < j) {
      swap(arr, i, j)
      i = i + 1
      j = j - 1
    }
    arr
  }
  
  def swap(arr: Array[Char], i: Int, j: Int): Unit = {
    val n = arr(i)
    val m = arr(j)
    val tmp = m ^ n
    arr(i) = (n ^ tmp).toChar
    arr(j) = (m ^ tmp).toChar
  }
  
  def main(args: Array[String]): Unit = {
    println(reverse(-2147483648))
    println(reverse(-2147483647))
    println(reverse(123))
    println(reverse(6))
    println(reverse(210))
    println(reverse(-123))
  }
}

File: src/main/java/com/leecode/easy/ReverseInteger.java
package com.leecode.easy;

public class ReverseInteger {
    public int reverse(int x) {
        if (x == Integer.MIN_VALUE) {
            return 0;
        }
        
        int res = 0;
        int n = x;
        while (n != 0) {
            int last = n % 10;
            if (last > 0 && res > (Integer.MAX_VALUE - last) / 10) {
                return 0;
            }
            if (last < 0 && res < (Integer.MIN_VALUE - last) / 10) {
                return 0;
            }
            res = res * 10 + last;
            n /= 10;
        }
        return res;
    }
    
    public int reverse2(int x) {
        String s = String.valueOf(x);
        char[] chars = s.toCharArray();
        char[] reversedChars = reverseChars(chars);
        try {
            return Integer.valueOf(String.valueOf(reversedChars));
        } catch (Exception e) {
            return 0;
        }
    }
    
    private char[] reverseChars(char[] chars) {
        if (chars[0] == '-') {
            return reverseChars(chars, 1, chars.length - 1);
        } else {
            return reverseChars(chars, 0, chars.length - 1);
        }
    }
    
    private char[] reverseChars(char[] chars, int start, int end) {
        if (end - start == 0) {
            return chars;
        }
        int s = start;
        int e = end;
        while (s < e) {
            swap(chars, s, e);
            ++s;
            --e;
        }
        return chars;
    }
    
    private void swap(char[] chars, int s, int e) {
        char t = chars[s];
        chars[s] = chars[e];
        chars[e] = t;
    }
    
    public static void main(String[] args) {
        ReverseInteger app = new ReverseInteger();
        System.out.println(app.reverse(1534236469));
        System.out.println(app.reverse(123));
        System.out.println(app.reverse(-123));
        System.out.println(app.reverse(120));
        System.out.println(app.reverse(3));
        System.out.println(app.reverse(45));
    }
}